src/main/java/org/broadinstitute/hellbender/tools/coveragemodel/CoverageModelEMAlgorithm.java

```diff
@@ -0,0 +1,497 @@
+            }
+        }
+
+        showIterationHeader();
+
+        double prevEStepLikelihood = Double.NEGATIVE_INFINITY;
+        double latestEStepLikelihood = Double.NEGATIVE_INFINITY;
+        double prevMStepLikelihood = Double.NEGATIVE_INFINITY;
+        double latestMStepLikelihood = Double.NEGATIVE_INFINITY;
+        final EMAlgorithmIterationInfo iterInfo = new EMAlgorithmIterationInfo(Double.NEGATIVE_INFINITY, 0, 1);
+        boolean updateCopyRatioPosteriors = false;
+        boolean paramEstimationConverged = false;
+        boolean performMStep = true;
+
+        while (iterInfo.iter <= params.getMaxEMIterations()) {
+
+            /* cycle through E-step mean-field equations until they are satisfied to the desired degree */
+            double maxPosteriorErrorNorm = 0;
+            int iterEStep = 0;
+
+            while (iterEStep < params.getMaxEStepCycles()) {
+                double posteriorErrorNormReadDepth, posteriorErrorNormSampleUnexplainedVariance,
+                        posteriorErrorNormBias, posteriorErrorNormCopyRatio;
+
+                runRoutine(this::updateReadDepthLatentPosteriorExpectations, s -> "N/A", "E_STEP_D", iterInfo);
+                posteriorErrorNormReadDepth = iterInfo.errorNorm;
+
+                runRoutine(this::updateBiasLatentPosteriorExpectations, s -> "N/A", "E_STEP_Z", iterInfo);
+                posteriorErrorNormBias = iterInfo.errorNorm;
+
+                if (params.gammaUpdateEnabled()) {
+                    runRoutine(this::updateSampleUnexplainedVariance,
+                            s -> "iters: " + s.getInteger("iterations"), "E_STEP_GAMMA", iterInfo);
+                    posteriorErrorNormSampleUnexplainedVariance = iterInfo.errorNorm;
+                } else {
+                    posteriorErrorNormSampleUnexplainedVariance = 0;
+                }
+
+                if (updateCopyRatioPosteriors) {
+                    runRoutine(this::updateCopyRatioLatentPosteriorExpectations, s -> "N/A", "E_STEP_C", iterInfo);
+                    posteriorErrorNormCopyRatio = iterInfo.errorNorm;
+                } else {
+                    posteriorErrorNormCopyRatio = 0;
+                }
+
+                /* calculate the maximum change of posteriors in this cycle */
+                maxPosteriorErrorNorm = Collections.max(Arrays.asList(
+                        posteriorErrorNormReadDepth,
+                        posteriorErrorNormSampleUnexplainedVariance,
+                        posteriorErrorNormBias,
+                        posteriorErrorNormCopyRatio));
+
+                /* check convergence of the E-step */
+                if (maxPosteriorErrorNorm < params.getPosteriorAbsTol()) {
+                    break;
+                }
+
+                iterEStep++;
+            }
+
+            if (maxPosteriorErrorNorm > params.getPosteriorAbsTol()) {
+                logger.warn("E-step cycles did not fully converge. Increase the maximum number of E-step cycles." +
```

Comment:
Convergence of every E-step is not required for the convergence of the EM algorithm. The exit signal of `runExpectation` and `runExpectationMaximization` tells whether convergence is finally achieved (according to the given tolerances).